File: vargplvm/matlab/linard2biasVardistPsi2Gradient.m
function [gKern1, gKern2, gVarmeans, gVarcovars, gInd] = linard2biasVardistPsi2Gradient(linardKern, biasKern, vardist, Z, covGrad, learnInducing)

% LINARD2BIASVARDISTPSI2GRADIENT description.
  
% VARGPLVM
  
if nargin < 6
    learnInducing = 1;
end

% variational means
N = size(vardist.means,1);
%  inducing variables 
[M Q] = size(Z); 

[Psi2, Pnobias, Psi1] = linard2biasVardistPsi2Compute(linardKern, biasKern, vardist, Z);


% inverse variances
A = linardKern.inputScales;

% gradient for the bias parameter  
gKern2 = sum(sum(Pnobias.*covGrad)); 
Bnm = biasKern.variance*ones(size(Psi1)); 
BPsi1Covg = (Bnm*covGrad);

gInd = zeros(M, Q);

for q=1:vardist.latentDimension
   % 
   gKern(q) = sum(sum((vardist.means(:,q)*(Z(:,q)')).*BPsi1Covg));
   
   gVarmeans(:,q) = A(q)*sum((ones(vardist.numData,1)*Z(:,q)').*BPsi1Covg,2);
   if learnInducing
        gInd(:,q) = A(q)*sum((vardist.means(:,q)*ones(1,size(Z,1))).*BPsi1Covg,1)';
   end
   %   
end
%

gKern1 = 2*gKern(:)';  
% gVarmeans is N x Q matrix (N:number of data, Q:latent dimension)
% this will unfold this matrix column-wise 
gVarmeans = 2*gVarmeans(:)'; 

% gInd is M x Q matrix (M:number of inducing variables, Q:latent dimension)
% this will unfold this matrix column-wise 
gInd = 2*gInd(:)'; 

gVarcovars = zeros(1,prod(size(vardist.covars))); 

File: R/R/linard2biasVardistPsi2Gradient.R
linard2biasVardistPsi2Gradient <- function (linardKern, biasKern, vardist, Z, covGrad)
{
# % LINARD2BIASVARDISTPSI2GRADIENT description not available.  
# % FORMAT
# % DESC 
# % description not available
  
# variational means
  N <- dim(vardist$means)[1] 
# inducing variables 
  M <- dim(Z)[1]  
  Q <- dim(Z)[2]
  
  #[Psi2, Pnobias, Psi1]
  l2bvp2c<- linard2biasVardistPsi2Compute(linardKern, biasKern, vardist, Z) 
  Psi2 <- l2bvp2c$Psi2
  Pnobias <- l2bvp2c$Pnobias
  Psi1 <- l2bvp2c$Psi1
  
# % inverse variances
  A <- linardKern$inputScales 
  
# % gradient for the bias parameter  
  gKern2 <- sum(Pnobias*covGrad) 
  Bnm <- biasKern$variance*matrix(1, dim(Psi1)[1], dim(Psi1)[2])  
  BPsi1Covg <- (Bnm%*%covGrad) 
  
  gKern <- rep(0, vardist$latentDimension)
  gVarmeans <- matrix(0, vardist$numData, vardist$latentDimension)
  gInd <- matrix(0, dim(Z)[1], vardist$latentDimension)
  for (q in 1:vardist$latentDimension)
  {
    gKern[q] <- sum((vardist$means[,q]%*%t(Z[,q]))*BPsi1Covg)
    
    gVarmeans[,q] <- A[q]*rowSums((matrix(1, vardist$numData,1)%*%t(Z[,q]))*BPsi1Covg) 
    gInd[,q] <- A[q]*t(colSums((vardist$means[,q]%*%matrix(1, 1,dim(Z)[1]))*BPsi1Covg)) 
    
  }
  
  gKern1 <- 2*matrix(gKern, nrow = 1)  
# % gVarmeans is N x Q matrix (N:number of data, Q:latent dimension)
# % this will unfold this matrix column-wise 
  gVarmeans <- 2*matrix(gVarmeans, nrow = 1) 
  
# % gInd is M x Q matrix (M:number of inducing variables, Q:latent dimension)
# % this will unfold this matrix column-wise 
  gInd <- 2*matrix(gInd, nrow = 1) 
  
  gVarcovars <- matrix(0, 1,prod(dim(vardist$covars)))
  
  return (list(gKern1 = gKern1, gKern2 = gKern2, gVarmeans = gVarmeans, 
               gVarcovars = gVarcovars, gInd = gInd))
}
